src/backend/distributed/commands/multi_copy.c

```diff
@@ -2480,6 +2484,12 @@ CitusSendTupleToPlacements(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest
 
 			/* additionaly, we need to send the current tuple too */
 			sendTupleOverConnection = true;
+
+			/* make sure that we keep track of relation accesses */
+			placement =
+				SearchShardPlacementInList(placementList, connection->hostname,
+										   connection->port, false);
+			AssociatePlacementAccessWithRelation(placement, PLACEMENT_ACCESS_DML);
```

Comment:
This function is actually kind of dumb. It does the same thing for every placement in a relation. Can we just do this once for the whole statement?

Reply:
We do it for parallel queries via `RecordParallelRelationAccessForTaskList()`. I think we don't like that approach much as well because that requires any code-refactoring to keep that in mind explicitly calling the function. Could be an option with some larger refactoring that touches router/real-time executors as well.

Besides, I'm thinking that #2867 is the complementary change for this. With that, (a) we're already guarding against disasters like this -- I might consider adding more comments (b) giving a chance to have the infrastructure to record all relation accesses pretty easily in case any future change needs it. Obviously we should improve the performance first.

Reply:
ok, but why would we do it per placement in the COPY logic instead of per relation in the COPY logic. Wouldn't the latter be less code to maintain?

The naming and signature of `AssociatePlacementAccessWithRelation` also seems quite misleading, which is likely what led to this bug.

Reply:
> ok, but why would we do it per placement in the COPY logic instead of per relation in the COPY logic. Wouldn't the latter be less code to maintain?

Makes sense.

> The naming and signature of AssociatePlacementAccessWithRelation also seems quite misleading, which is likely what led to this bug.

That's also true. I think the intention for fkey to reference tables was to only record it for reference tables. But since we implemented the relation accesses as a separate infrastructure, we ended up with.